### ISSUE TITLE
Add JSDoc documentation to Webpack 2

### ISSUE THREAD

**Issue by kevinlaw91** [webpack/webpack#2969]
**I'm submitting a feature request**

Please consider adding [JSDoc](http://usejsdoc.org/) documentation to Webpack 2. This will allow code completion/inspection of webpack config files for IDE that supports it, such as WebStorm.

**Webpack version:**
2.x

**Comment by gajus** [comment]
I would much rather see a Flow type annotations than JSDoc.

JSDoc gets easily de-synced with the code. It does not provide any use outside of documentation.

Flow type does not need to enforced. It can be used to simply document the code. Here is an example of a codebase that uses Flow type annotations to document the code.

https://github.com/gajus/gitinfo/blob/master/src/gitinfo.js

**Comment by TheLarkInn** [comment]
Or Typescript 😏.

**Comment by TheLarkInn** [comment]
@kevinlaw91 this would be a substantial undertaking, but also incredibly impactful. We would certainly consider a PR.

**Comment by gajus** [comment]
> Or Typescript 😏.

I am not proposing to adapt strict typing. As you said, that would be a significant chunk of work.

I am suggesting to use types for documentation purposes instead of JSDoc, e.g.

Instead of:

```js
/**
 * @param {string} value The haystack value.
 * @param {string} searchValue The needle value.
 * @returns {boolean}
 */
const includes = (value, searchValue) => {
	
};
```

Use:

```js
/**
 * @param value The haystack value.
 * @param searchValue The needle value.
 */
const includes = (value: string, searchValue: string): boolean => {
	
};
```

it becomes even more succinct if you don't use description:

```js
const includes = (value: string, searchValue: string): boolean => {
	
};
```

You can then leverage https://github.com/gcanti/babel-plugin-tcomb or https://github.com/codemix/babel-plugin-typecheck to enable type checking during runtime (for testing purposes).

This does not require full transformation/ compliance to Flow or TS.

**Comment by kevinlaw91** [comment]
I believe converting to strict typing would be a different topic here. The main idea is to start adding basic documentation within the code.

Flow is nice to have, I do agree with @gajus . However, consider the effort needed for the transformation, it only brings huge benefits to webpack developers and offer less to the normal webpack users.

JSDoc are easier and faster to implement. and both users and developers can enjoy the advantages.
- Type definition is optional in JSDoc, much easier to write
- Javadoc like syntax, don't have to learn Flow to write
- Selectively documents parts of the code (most important first) and cover more later
- In the future when the time is right, existing JSDoc can be transformed to Flow with less work too

**Comment by gajus** [comment]
> However, consider the effort needed for the transformation

The only addition would be a required to have a transpilation step to either remove Flow annotations or to convert them to JSDoc.

> Javadoc like syntax, don't have to learn Flow to write

This assumes that more developers are familiar with JSDoc syntax than with Flow. While that might be true, JSDoc is also infamous for inconsistent use. Hey, but either, I have got you covered regarding consistency:

* https://github.com/gajus/eslint-plugin-flowtype
* https://github.com/gajus/eslint-plugin-jsdoc

:-)

> Selectively documents parts of the code (most important first) and cover more later

That is equally true to Flow and JSDoc.

> In the future when the time is right, existing JSDoc can be transformed to Flow with less work too

It is easier to transform Flow to JSDoc than JSDoc to Flow.

**Comment by leighlawhon** [comment]
I think "code-only documentation" or **"C.O.D (piece)"** as I like to call it, is great for experienced developers. However, I use documentation in my workflow specifically because we don't always live in a perfect world with top notch developers writing every line of code with the ability to understand why a linter is complaining or why the world works the way it does. There is more than one way to work and I, preferably, like to make sure that the tools I choose, like webpack, are inclusive to all levels of developers and even non-developers who sometimes need to understand your work. 

TL;DR:  +1 for jsDocs. If you don't like it, don't use it.

**Comment by gajus** [comment]
> TL;DR: 👍 for jsDocs. If you don't like it, don't use it. :-)

Is not an argument. It creates additional work and requires to maintain duplicate set of information, which increases the odds of an error.

**Comment by leighlawhon** [comment]
> Is not an argument. 

Correct, it's a discuss...and just adding a point of view and support, which is "having jsDoc in Webpack would be nice-to-have **inclusive** feature and it doesn't prevent you (or I for that matter) from using Flow."

**Comment by TheLarkInn** [comment]
I would love to implement JSDOC for webpack. Our only hurdle is that our code is run as is via node. That is to say that we do not transpile our SRC when the user uses it. So although I see the obnoxious benefits, we can't jump to it just yet without already pursuing some sort of build process which doesn't exist today.

**Comment by leighlawhon** [comment]
> obnoxious benefits,
 I think you mean "obvious" but obnoxious works too
